controllers/scenario/scenario_adapter.go

```diff
@@ -0,0 +1,122 @@
+*/
+
+package scenario
+
+import (
+	"context"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/go-logr/logr"
+	applicationapiv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/gitops"
+	"github.com/redhat-appstudio/operator-goodies/reconciler"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Adapter holds the objects needed to reconcile a Release.
+type Adapter struct {
+	application *applicationapiv1alpha1.Application
+	scenario    *v1alpha1.IntegrationTestScenario
+	logger      logr.Logger
+	client      client.Client
+	context     context.Context
+}
+
+// NewAdapter creates and returns an Adapter instance.
+func NewAdapter(application *applicationapiv1alpha1.Application, scenario *v1alpha1.IntegrationTestScenario, logger logr.Logger, client client.Client,
+	context context.Context) *Adapter {
+	return &Adapter{
+		application: application,
+		scenario:    scenario,
+		logger:      logger,
+		client:      client,
+		context:     context,
+	}
+}
+
+// EnsureCreatedScenarioIsValid is an operation that ensures created IntegrationTestScenario is valid
+// in case it is, set its owner reference
+func (a *Adapter) EnsureCreatedScenarioIsValid() (reconciler.OperationResult, error) {
+
+	// first check if application exists or not
+	if a.application == nil {
+		a.logger.Info("Application for scenario was not found.",
+			"Sceario.Name", a.scenario.Name)
+
+		patch := client.MergeFrom(a.scenario.DeepCopy())
+		SetScenarioIntegrationStatusAsInvalid(a.scenario, "Failed to get application for scenario.")
+		return reconciler.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.scenario, patch))
+		// if application exists, check if scenario has ownerReference assigned
+	} else if a.scenario.OwnerReferences == nil {
+		patch := client.MergeFrom(a.scenario.DeepCopy())
+		err := ctrl.SetControllerReference(a.application, a.scenario, a.client.Scheme())
+		if err != nil {
+			a.logger.Error(err, "Error setting owner reference.",
+				"Application.Name", a.application.Name,
+				"Application.Namespace", a.application.Namespace)
+			patch := client.MergeFrom(a.scenario.DeepCopy())
+			SetScenarioIntegrationStatusAsInvalid(a.scenario, "Failed to get application for scenario.")
+			return reconciler.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.scenario, patch))
+		}
+		err = a.client.Patch(a.context, a.scenario, patch)
+		if err != nil {
+			a.logger.Error(err, "Failed to update Scenario",
+				"Scenario.Name", a.scenario.Name)
+			return reconciler.RequeueWithError(err)
+		}
+		return reconciler.ContinueProcessing()
+		// app exists, owner reference is set, check if the scenario has envirnment defined
+	} else if reflect.ValueOf(a.scenario.Spec.Environment).IsZero() {
```

Comment:
I think we'll need to decouple these `if` conditions because we want to check the scenario for multiple things in the same reconcile pass.

Reply:
Decoupled if statements.